packages/core/src/verification/signature-verifier.ts

```diff
@@ -0,0 +1,543 @@
+        this.logger.debug('Error fetching signature data', {
+          error: error.message,
+          stack: error.stack,
+          txId,
+          gateway: gateway.toString(),
+        });
+        // Continue to the next gateway if there's an error
+        continue;
+      }
+    }
+
+    throw new Error('Failed to fetch signature data from any trusted gateway', {
+      cause: { txId },
+    });
+  }
+
+  async verifyData({
+    data,
+    txId,
+  }: { data: DataStream; txId: string }): Promise<void> {
+    const {
+      format,
+      owner,
+      target,
+      quantity,
+      reward,
+      dataSize,
+      lastTx,
+      tags,
+      signature,
+    } = await this.getSignatureData({ txId });
+
+    const tagList: [Uint8Array, Uint8Array][] = tags.map(
+      (tag: { name: string; value: string }) => [
+        fromB64Url(tag.name),
+        fromB64Url(tag.value),
+      ],
+    );
+
+    // use the provided data stream to compute the data root for the signature data
+    const computedDataRoot = await convertDataStreamToDataRoot({
+      stream: data,
+    });
+
+    // compute the signature data using the computed data root and retrieved signature data
+    const signatureData = await deepHash([
+      stringToBuffer(format.toString()),
+      fromB64Url(owner),
+      fromB64Url(target),
+      stringToBuffer(quantity),
+      stringToBuffer(reward),
+      fromB64Url(lastTx),
+      tagList,
+      stringToBuffer(dataSize.toString()),
+      fromB64Url(computedDataRoot),
+    ]);
+
+    // verify the signature using the computed signature data and the computed signature data
+    const isValid = await Arweave.crypto.verify(
+      owner,
+      signatureData,
+      fromB64Url(signature),
+    );
+    if (!isValid) {
+      throw new Error('Transaction signature verification failed', {
+        cause: { txId },
+      });
+    }
+  }
+}
+
+export const SignatureVerificationStrategies = {
+  ans104: Ans104SignatureVerificationStrategy,
+  transaction: TransactionSignatureVerificationStrategy,
+  // TODO: ans102
+};
+
+export class SignatureVerificationStrategy {
+  private readonly ans104: Ans104SignatureVerificationStrategy;
+  private readonly transaction: TransactionSignatureVerificationStrategy;
+  private readonly classifier: DataClassifier;
+  constructor({
+    trustedGateways,
+    maxConcurrency = 1,
+    logger = defaultLogger,
+    classifier = new GqlClassifier({ logger }),
+  }: {
+    trustedGateways: URL[];
+    maxConcurrency?: number;
+    logger?: Logger;
+    classifier?: DataClassifier;
+  }) {
+    this.ans104 = new Ans104SignatureVerificationStrategy({
+      trustedGateways,
+      maxConcurrency,
+      logger,
+    });
+    this.transaction = new TransactionSignatureVerificationStrategy({
+      trustedGateways,
+      logger,
+    });
+    this.classifier = classifier;
+  }
+
+  async verifyData({
+    data,
+    txId,
+  }: { data: DataStream; txId: string }): Promise<void> {
```

Comment:
Nit: data is an overloaded enough term to potentially justify a rename here to `dataStream`